config/cmake/ConfigureChecks.cmake

```diff
@@ -923,46 +923,56 @@ if (HDF5_ENABLE_NONSTANDARD_FEATURE_FLOAT16)
       # compile a program that will generate these functions to check for _Float16
       # support. If we fail to compile this program, we will simply disable
       # _Float16 support for the time being.
+      H5ConversionTests (
+          ${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK
+          FALSE
+          "Checking if compiler can convert _Float16 type with casts"
+      )
 
       # Some compilers, notably AppleClang on MacOS 12, will succeed in the
-      # configure check below when optimization flags like -O3 are manually
+      # configure check above when optimization flags like -O3 are manually
       # passed in CMAKE_C_FLAGS. However, the build will then fail when it
       # reaches compilation of H5Tconv.c because of the issue mentioned above.
-      # MacOS 13 appears to have fixed this, but, just to be sure, backup and
-      # clear CMAKE_C_FLAGS before performing these configure checks.
+      # MacOS 13 appears to have fixed this, but, just to be sure, make sure
+      # the check also passes without the passed in CMAKE_C_FLAGS.
       set (cmake_c_flags_backup "${CMAKE_C_FLAGS}")
       set (CMAKE_C_FLAGS "")
 
       H5ConversionTests (
-          ${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK
+          ${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK_NO_FLAGS
           FALSE
-          "Checking if compiler can convert _Float16 type with casts"
+          "Checking if compiler can convert _Float16 type with casts (without CMAKE_C_FLAGS)"
       )
 
       set (CMAKE_C_FLAGS "${cmake_c_flags_backup}")
 
-      if (${${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK})
+      if (${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK AND ${HDF_PREFIX}_FLOAT16_CONVERSION_FUNCS_LINK_NO_FLAGS)
         # Finally, MacOS 13 appears to have a bug specifically when converting
```

Comment:
Is the issue related to the OS or the C clang version? Though I've not tried it, I would assume that gcc installed on 13 OS systems would work.

Reply:
I'm just referring to the default AppleClang here for a particular MacOS version. I'd assume GCC should work as well, but haven't tested that.